src/Sarif/Query/ExpressionParser.cs

```diff
@@ -0,0 +1,354 @@
+                    terms.Add(ParseAndExpression(ref text));
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            return (terms.Count == 1 ? terms[0] : new OrExpression(terms));
+        }
+
+        private static IExpression ParseAndExpression(ref StringSlice text)
+        {
+            List<IExpression> terms = new List<IExpression>();
+            terms.Add(ParseTerm(ref text));
+
+            while (true)
+            {
+                Literal<ExpressionToken> t = StartingToken(ref text);
+                if (t?.Value == ExpressionToken.And)
+                {
+                    text = text.Substring(t.Text.Length);
+                    terms.Add(ParseTerm(ref text));
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            return (terms.Count == 1 ? terms[0] : new AndExpression(terms));
+        }
+
+        private static IExpression ParseTerm(ref StringSlice text)
+        {
+            Literal<ExpressionToken> t = StartingToken(ref text);
+
+            // All?
+            if (t?.Value == ExpressionToken.All)
+            {
+                text = text.Substring(t.Text.Length);
+                return new AllExpression();
+            }
+
+            // Parenthesized subexpression?
+            if (t?.Value == ExpressionToken.LeftParen)
+            {
+                text = text.Substring(t.Text.Length);
+                IExpression subexpression = ParseExpression(ref text);
+
+                t = StartingToken(ref text);
+                if (t?.Value != ExpressionToken.RightParen) { throw new QueryParseException("nested expression end paren", text); }
+                text = text.Substring(t.Text.Length);
+
+                return subexpression;
+            }
+
+            // Not?
+            if (t?.Value == ExpressionToken.Not)
+            {
+                text = text.Substring(t.Text.Length);
+                return new NotExpression(ParseTerm(ref text));
+            }
+
+            // PropertyName CompareOperator Value
+            StringSlice propertyName = ParseString(ref text);
+            CompareOperator op = ParseCompareOperator(ref text);
+            StringSlice value = ParseString(ref text);
+
+            return new TermExpression(propertyName.ToString(), op, value.ToString());
+        }
+
+        private static CompareOperator ParseCompareOperator(ref StringSlice text)
+        {
+            ConsumeWhitespace(ref text);
+
+            foreach (Literal<CompareOperator> op in CompareOperators)
+            {
+                if (text.StartsWith(op.Text, StringComparison.OrdinalIgnoreCase))
+                {
+                    text = text.Substring(op.Text.Length);
+                    return op.Value;
+                }
+            }
+
+            throw new QueryParseException("compare operator", text);
```

Comment:
>text [](start = 62, length = 4)

Should the exception print the StringSlice as it entered the method? i.e. it looks like this method modifies the StringSlice, so it's unclear if we want the original text or the updated text in the exception.

Reply:
The exception should write the slice, which should start with the thing the parser had trouble with. Each place in the parser which throws should not consume the objectionable thing. In this case, we only throw if none of the operators in the loop were recognized, so whatever was first will not have been consumed.